kamu-core/src/infra/repos/dataset_repository_s3.rs

```diff
@@ -0,0 +1,473 @@
+    endpoint: String,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+impl DatasetRepositoryS3 {
+    pub fn new(s3_context: S3Context, endpoint: String) -> Self {
+        Self {
+            s3_context,
+            endpoint,
+        }
+    }
+
+    fn get_s3_bucket_path(&self, dataset_name: &str) -> Url {
+        let dataset_url_string = format!(
+            "s3+{}/{}/{}/",
+            self.endpoint,
+            self.s3_context.bucket,
+            self.s3_context.get_key(dataset_name)
+        );
+        Url::parse(dataset_url_string.as_str()).unwrap()
+    }
+
+    fn get_dataset_impl(&self, dataset_name: &DatasetName) -> Result<impl Dataset, InternalError> {
+        let dataset_url = self.get_s3_bucket_path(dataset_name.as_str());
+        DatasetFactoryImpl::get_s3(dataset_url)
+    }
+
+    async fn eliminate_bucket_items(
+        &self,
+        dataset_name: &DatasetName,
+    ) -> Result<(), InternalError> {
+        let dataset_key_prefix = self.s3_context.get_key(dataset_name.as_str());
+        let list_response = self
+            .s3_context
+            .client
+            .list_objects(ListObjectsRequest {
+                bucket: self.s3_context.bucket.clone(),
+                prefix: Some(dataset_key_prefix),
+                ..ListObjectsRequest::default()
+            })
+            .await
+            .int_err()?;
+
+        if let Some(contents) = list_response.contents {
+            self.s3_context
+                .client
+                .delete_objects(DeleteObjectsRequest {
+                    bucket: self.s3_context.bucket.clone(),
+                    delete: Delete {
+                        objects: contents
+                            .iter()
+                            .map(|obj| ObjectIdentifier {
+                                key: obj.key.clone().unwrap(),
+                                version_id: None,
+                            })
+                            .collect(),
+                        quiet: Some(true),
+                    },
+                    ..DeleteObjectsRequest::default()
+                })
+                .await
+                .int_err()?;
+        }
+
+        Ok(())
+    }
+
+    async fn finish_create_dataset(
+        &self,
+        dataset: &dyn Dataset,
+        staging_name: &str,
+        dataset_name: &DatasetName,
+    ) -> Result<DatasetHandle, CreateDatasetError> {
+        let summary = match dataset.get_summary(GetSummaryOpts::default()).await {
+            Ok(s) => Ok(s),
+            Err(GetSummaryError::EmptyDataset) => unreachable!(),
+            Err(GetSummaryError::Access(e)) => Err(e.int_err().into()),
+            Err(GetSummaryError::Internal(e)) => Err(CreateDatasetError::Internal(e)),
+        }?;
+
+        let handle = DatasetHandle::new(summary.id, dataset_name.clone());
+
+        match self
+            .move_bucket_items_on_dataset_rename(staging_name, dataset_name.as_str())
```

Comment:
Moving in S3 is expensive, so since it doesn't give us atomicity mb we shouldn't even bother, and better solve the concurrency problem on the database level.